posts/2017-12-20-who-implements-typeclass.md

```diff
@@ -0,0 +1,565 @@
+    | +--------+            | |                    +===========+
+    | | zipAdd |---→---→---→- |
+    | +--------+    calls     |  +------+ +---------+ +-------+
+    ↑        ↑                |  | Ints | | Strings | | Vects |
+    |   calls|                |  +------+ +---------+ +-------+
+    |      +-----+            |      |
+    |      | ??? |            |      |
+    ↑      +-----+            |      |
+    |  (derived combinators          ↓
+    |   can derive from each other)  ---→---→---→
+    |                                           |
+    ↑          -------------------------------  |
+    |          To evaluate `addThree(1, 2, 3)`  |
+    |          -------------------------------  ↓
+    |          1. Fetch `Adder` implicitly      |
+    |-←---←---←---←---←---←---←---←---←---←---←-|
+               2. Pass to `addThree`
+               3. `addThree` uses the abstract interface to
+                  invoke the primitive `add` combinator on what,
+                  to it, is an abstract type, `A`.
+```
+
+## Infinite overloads via recursion
+
+Making derived combinators easier to write is very useful, but
+typeclasses go further by letting you describe overloading rules that
+would be impossible with normal overloading.
+
+Given that I can add `Int` and `Int` together, I should be able to add
+`(Int, Int)` and `(Int, Int)` to get `(Int, Int)`.
+
+```scala
+implicit val addIntPairs: Adder[(Int, Int)] =
+  Adder{case ((x1, x2), (y1, y2)) =>
+    (x1 + y1, x2 + y2)}
+    
+scala> add((2, 7), (3, 8))
+res3: (Int, Int) = (5,15)
+```
+
+But I should also be able to add pairs of `String`. And `(Int,
+String)` pairs. And `(String, Vector[Boolean])` pairs. And pairs of
+pairs of pairs.
+
+Typeclasses let you declare newly supported types recursively, with an
+implicit argument list to the `implicit def`.
+
+```scala
+implicit def addPairs[A: Adder, B: Adder]
+    : Adder[(A, B)] =
+  Adder{case ((a1, b1), (a2, b2)) =>
+    (add(a1, a2), add(b1, b2))
+  }
+```
+
+## Surely this must be going somewhere new
+
+If you’re familiar with type classes, all this must be old hat. But
+this time, we’re going to expand the boundaries of the typeclass
+design space, by exploiting *GADT pattern matching*.
+
+We could have designed the `Adder` type class to include `addThree` as
+a primitive combinator, and implemented it afresh for each of the four
+instances we’ve defined so far, as well as any future instances
+someone might define. Thinking orthogonally, however, shows us that
+there’s a more primitive concept which strictly generalizes it: if we
+primitively define a two-value adder, we can use it to add three
+items, simply by using it twice.
+
+This has a direct impact on how we structure the functions related to
+`Adder`. The primitives must be split up, their separate
+implementations appearing directly in the implicit instances. Derived
+combinators may occur anywhere that is convenient to us.
```

Comment:
This is a trade-off: you might want to include more derived combinators in the classes so that instances can override them for performance.

Reply:
There's even more to say here, which I've added in 5212fe4; thanks again.